### PR TITLE
build docker from makefile

### PR DESCRIPTION
**Pull request by thedavidmeister**
child of #309 

this is prework for moving travis builds into native travis instead of docker (massive speedup)

main goals:

- _always_ use an explicit toolchain when interacting with cargo/rustup in the `Makefile`
- have `Dockerfile.ubuntu` defer to the `Makefile` for the build, to help enforce versions
- `Makefile` sets default versions for core and tooling
- wasm no longer treated as separate case for versioning (rolled into core)

this is important because the Travis cache is very sensitive to being completely invalidated by any new nightly versions landing and being incidentally installed, even if we don't use those versions

installation via. the makefile is idempotent because outside docker we cannot control what is already on the host system

using `if` rather than `||` for the idempotent controls because `||` only responds to the existence of an error. `||` behaves the same for something already installed as something not being able to be installed.

travis tests probably won't help much here because there's a new docker build

new docker builds coming through docker hub: https://hub.docker.com/r/holochain/holochain-rust/builds/

Makefile symlink in `/docker` is for docker hub (which can't "see" outside the `Dockerfile` dir)

## followup

- tests not being run #318

### CODE REVIEW

File: docker/Makefile [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+	qmake -o $@Makefile $@qmake.pro
+	cd $@; $(MAKE)
+
+# execute all tests, both rust and "C" bindings
+test: test_non_c c_binding_tests ${C_BINDING_TESTS}
+
+test_non_c: main
+	RUSTFLAGS="-D warnings" $(CARGO) test
+
+test_c_ci: c_binding_tests ${C_BINDING_TESTS}
```
Comment (lucksus) [inline docker/Makefile]: This needs to also build and run `core_api_c_binding`.
See my comments in you other travis PR #309.

Comment (lucksus) [reply docker/Makefile]: Hm, could it be that we never tested `core_api_c_binding`?

**Comment by lucksus** [comment]
local `. docker/run-test`

```
test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

   Doc-tests test_utils

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

qmake -o c_binding_tests/hc_dna/Makefile c_binding_tests/hc_dna/qmake.pro
cd c_binding_tests/hc_dna/; make
make[1]: Entering directory '/holochain/c_binding_tests/hc_dna'
g++ -Wl,-O1 -o ../../target/debug/c_binding_tests/hc_dna/test_executable ../../target/debug/c_binding_tests/hc_dna/.obj/test.o ../../target/debug/c_binding_tests/hc_dna/.obj/moc_test.o   -L../../target/debug/ -lholochain_dna_c_binding -ldl -L/usr/lib/x86_64-linux-gnu -lQt5Test -lQt5Core -lpthread 
../../target/debug/c_binding_tests/hc_dna/.obj/test.o: file not recognized: File format not recognized
collect2: error: ld returned 1 exit status
Makefile:138: recipe for target '../../target/debug/c_binding_tests/hc_dna/test_executable' failed
make[1]: *** [../../target/debug/c_binding_tests/hc_dna/test_executable] Error 1
make[1]: Leaving directory '/holochain/c_binding_tests/hc_dna'
Makefile:86: recipe for target 'c_binding_tests/hc_dna/' failed
make: *** [c_binding_tests/hc_dna/] Error 2
lucksus at lucksus15 in ~/code/holochain-rust on 305-travis-raw-version-review*
$ 
```

**Comment by thedavidmeister** [comment]
@lucksus 

```
qmake -o c_binding_tests/hc_dna/Makefile c_binding_tests/hc_dna/qmake.pro
cd c_binding_tests/hc_dna/; make
make[1]: Entering directory '/holochain/c_binding_tests/hc_dna'
g++ -c -pipe -O2 -Wall -W -D_REENTRANT -fPIC -DQT_NO_DEBUG -DQT_TESTLIB_LIB -DQT_CORE_LIB -DQT_TESTCASE_BUILDDIR='"/holochain/c_binding_tests/hc_dna"' -I. -I. -isystem /usr/include/x86_64-linux-gnu/qt5 -isystem /usr/include/x86_64-linux-gnu/qt5/QtTest -isystem /usr/include/x86_64-linux-gnu/qt5/QtCore -I../../target/debug/c_binding_tests/hc_dna/.moc -I/usr/lib/x86_64-linux-gnu/qt5/mkspecs/linux-g++ -o ../../target/debug/c_binding_tests/hc_dna/.obj/test.o test.cpp
g++ -pipe -O2 -Wall -W -dM -E -o ../../target/debug/c_binding_tests/hc_dna/.moc/moc_predefs.h /usr/lib/x86_64-linux-gnu/qt5/mkspecs/features/data/dummy.cpp
/usr/lib/qt5/bin/moc -DQT_NO_DEBUG -DQT_TESTLIB_LIB -DQT_CORE_LIB -DQT_TESTCASE_BUILDDIR='"/holochain/c_binding_tests/hc_dna"' --include ../../target/debug/c_binding_tests/hc_dna/.moc/moc_predefs.h -I/usr/lib/x86_64-linux-gnu/qt5/mkspecs/linux-g++ -I/holochain/c_binding_tests/hc_dna -I/holochain/c_binding_tests/hc_dna -I/usr/include/x86_64-linux-gnu/qt5 -I/usr/include/x86_64-linux-gnu/qt5/QtTest -I/usr/include/x86_64-linux-gnu/qt5/QtCore -I/usr/include/c++/7 -I/usr/include/x86_64-linux-gnu/c++/7 -I/usr/include/c++/7/backward -I/usr/lib/gcc/x86_64-linux-gnu/7/include -I/usr/local/include -I/usr/lib/gcc/x86_64-linux-gnu/7/include-fixed -I/usr/include/x86_64-linux-gnu -I/usr/include test.h -o ../../target/debug/c_binding_tests/hc_dna/.moc/moc_test.cpp
g++ -c -pipe -O2 -Wall -W -D_REENTRANT -fPIC -DQT_NO_DEBUG -DQT_TESTLIB_LIB -DQT_CORE_LIB -DQT_TESTCASE_BUILDDIR='"/holochain/c_binding_tests/hc_dna"' -I. -I. -isystem /usr/include/x86_64-linux-gnu/qt5 -isystem /usr/include/x86_64-linux-gnu/qt5/QtTest -isystem /usr/include/x86_64-linux-gnu/qt5/QtCore -I../../target/debug/c_binding_tests/hc_dna/.moc -I/usr/lib/x86_64-linux-gnu/qt5/mkspecs/linux-g++ -o ../../target/debug/c_binding_tests/hc_dna/.obj/moc_test.o ../../target/debug/c_binding_tests/hc_dna/.moc/moc_test.cpp
g++ -Wl,-O1 -o ../../target/debug/c_binding_tests/hc_dna/test_executable ../../target/debug/c_binding_tests/hc_dna/.obj/test.o ../../target/debug/c_binding_tests/hc_dna/.obj/moc_test.o   -L../../target/debug/ -lholochain_dna_c_binding -ldl -lQt5Test -lQt5Core -lpthread 
make[1]: Leaving directory '/holochain/c_binding_tests/hc_dna'
target/debug/c_binding_tests/hc_dna/test_executable
********* Start testing of TestHcDna *********
Config: Using QtTest library 5.9.5, Qt 5.9.5 (x86_64-little_endian-lp64 shared (dynamic) release build; by GCC 7.3.0)
PASS   : TestHcDna::initTestCase()
PASS   : TestHcDna::serializeAndDeserialize()
PASS   : TestHcDna::canGetName()
PASS   : TestHcDna::canSetName()
PASS   : TestHcDna::canGetZomeNames()
PASS   : TestHcDna::canGetCapabilityNames()
PASS   : TestHcDna::canGetFunctionNames()
PASS   : TestHcDna::canGetFunctionParameters()
PASS   : TestHcDna::cleanupTestCase()
Totals: 9 passed, 0 failed, 0 skipped, 0 blacklisted, 22ms
********* Finished testing of TestHcDna *********

```

**Pull request merged**